glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+        // one(s) to check based on the label variant.
+        let (key_count, category_count) = count_keys_and_categories(meta, glean);
+        match label {
+            DynamicLabelType::Label(ref label) => {
+                record_error(
+                    glean,
+                    meta,
+                    ErrorType::InvalidLabel,
+                    format!("Invalid `DualLabeledCounter` label format: {label:?}"),
+                    None,
+                );
+                key = OTHER_LABEL;
+                category = OTHER_LABEL;
+            }
+            DynamicLabelType::KeyOnly(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::CategoryOnly(_) => {
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+            DynamicLabelType::KeyAndCategory(_) => {
+                if key_count >= MAX_LABELS || !label_is_valid(key, glean, meta) {
+                    key = OTHER_LABEL;
+                }
+                if category_count >= MAX_LABELS || !label_is_valid(category, glean, meta) {
+                    category = OTHER_LABEL;
+                }
+            }
+        }
+        combine_base_identifier_and_labels(base_identifier, key, category)
+    } else {
+        record_error(
+            glean,
+            meta,
+            ErrorType::InvalidLabel,
+            "Invalid `DualLabeledCounter` label format, unable to determine key and/or category",
+            None,
+        );
+        combine_base_identifier_and_labels(base_identifier, OTHER_LABEL, OTHER_LABEL)
+    }
+}
+
+fn label_is_valid(label: &str, glean: &Glean, meta: &CommonMetricDataInternal) -> bool {
+    if label.len() > MAX_LABEL_LENGTH {
+        let msg = format!(
+            "label length {} exceeds maximum of {}",
+            label.len(),
+            MAX_LABEL_LENGTH
+        );
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else if label.chars().any(|c| !c.is_ascii() || c.is_ascii_control()) {
+        let msg = format!("label must be printable ascii, got '{:?}'", label);
+        record_error(glean, meta, ErrorType::InvalidLabel, msg, None);
+        false
+    } else {
+        true
+    }
+}
+
+fn count_keys_and_categories(meta: &CommonMetricDataInternal, glean: &Glean) -> (usize, usize) {
+    let base_identifier = &meta.base_identifier();
+    let prefix = format!("{base_identifier}{RECORD_SEPARATOR}");
+    let mut seen_keys: Vec<String> = Vec::new();
+    let mut seen_categorires: Vec<String> = Vec::new();
+    let mut snapshotter = |mut metric_id: &[u8], _: &Metric| {
+        let mut metric_id_str = String::new();
+        _ = metric_id.read_to_string(&mut metric_id_str);
```

Comment:
You should be able to use this:

```suggestion
        let mut metric_id_str = str::from_utf8(metric_id);
```

or potentially even better:

```suggestion
	let metric_id_str = String::from_utf8_lossy(metric_id);
```

No allocation needed if this is utf-8 (which it should be, because that's what we put in).